keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+    - Building out libraries for code reuse
+    - Validating Kubernetes resource YAML files adhere to a policy in a CI/CD pipeline
+- Mutation support. While we believe this enhancement could be extended in the
+  future to support mutations, we believe it is best handled as a separate
+  enhancement. That said, we need to keep mutation in mind in this KEP to ensure
+  we design it in such a way that we don't obviously paint ourselves into a
+  corner where mutation would be difficult to introduce later.
+- Full feature parity with validating admission webhooks. For example, this
+  enhancement is not expected to ever support making requests to external
+  systems.
+- Replace the admission controllers compiled into the API server. 
+- Static or on-initialization specification of admission config. This is a
+  needed feature but should be solved in a general way and not in this KEP
+  (xref: https://github.com/kubernetes/enhancements/issues/1872).
+
+## Background
+
+This is not a new idea, Tristan Swadell (@TristonianJones) explored policy for
+Kubernetes using CEL with
+[cel-policy-templates-go](https://github.com/google/cel-policy-templates-go),
+and Jordan Liggitt (@liggitt) prototyped using CEL for in-process admission
+control in Kubernetes in 2020.
+
+[CRD Validation
+Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+were implemented as a more constrained subset of this problem and addressed how
+to integrate the Kubernetes type system with CEL.
+
+## Considerations
+
+### Admission Webhook Parity
+
+Users of the Kubernetes API are already familiar with
+ValidatingWebhookConfigurations. We should strive for consistency with this API
+unless there is a good reason to diverge from it.  As a concrete example, we
+should provide access to all the information that webhooks have access to (see
+[AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+and if we provide access to additional information we should extend
+AdmissionRequest to include it. (But we need to be careful not to make
+AdmissionRequest significantly larger as this will impact the
+performance/latency of existing webhooks not leveraging the additional
+information.  We should be careful about providing access to cross-object
+information like namespace objects to webhooks since they can be stale.)
+
+### Configurability
+
+Consider an admission rule that disallows requests based on a blocklist.
+
+While it is possible to inline a blocklist directly into a CEL expression as a
+data literal (`!(object.metadata.name in ['blocked1', 'blocked2'])`), this quickly
+becomes problematic:
+
+- Long blocklists become unwieldy quickly in CEL expressions
+- A blocklist per scope (e.g. namespace) is inconvenient express and maintain
+
+The need to configure admission rules is common enough (see below use cases)
```

Comment:
Another common way to support this is to perform secondary authorization checks to the kube-apiserver authorizer to allow the configuration to happen in an RBAC rule. The `use` verb for PSP or openshift SCC is an example of this.